README.md
# time-control

Simple command-line time logging utility.

## Build

    $ lein uberjar

## Usage

    $ java -jar time-control-0.1.0-standalone.jar [args]

## License

Copyright © 2021 Andrey Listopadov

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-03-27
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-03-27
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/time-control/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/time-control/compare/0.1.0...0.1.1


test/time_control/log_test.clj
(ns time-control.log-test
  (:require [time-control.log :as log :refer [*log-dir*]]
            [me.raynes.fs :as fs]
            [clojure.test :refer [deftest is testing use-fixtures]]))

(defn- wrap-log-dir [test]
  (binding [*log-dir* "test/data/time-log/"]
    (test)))

(defn- prepare-logs [test]
  (let [files (conj (map str
                         (repeat "01-01-2021-")
                         (range 10 0 -1)
                         (repeat ".time-log"))
                    "01-01-2021.time-log")]
    (fs/mkdirs *log-dir*)
    (doseq [file files]
      (spit (fs/file *log-dir* file) "[]"))
    (test)
    (doseq [file files]
      (fs/delete (fs/file *log-dir* file)))
    (fs/delete-dir *log-dir*)))

(use-fixtures :each
  wrap-log-dir
  prepare-logs)

(deftest log-files
  (testing "get last log file"
    (is (= (str (fs/file *log-dir* "01-01-2021-10.time-log"))
           (str (log/find-previous-log)))))

  (testing "create new log file"
    (is (= (str (fs/file *log-dir* "01-01-2021-11.time-log"))
           (str (log/create-new-log "01-01-2021"))))
    (is (= (str (fs/file *log-dir* "01-02-2021.time-log"))
           (str (log/create-new-log "01-02-2021"))))
    (is (= (str (fs/file *log-dir* "01-02-2021-1.time-log"))
           (str (log/create-new-log "01-02-2021")))))

  (testing "get last log for specific date"
    (is (= (str (fs/file *log-dir* "01-01-2021-11.time-log"))
           (str (log/find-previous-log "01-01-2021"))))
    (is (= (str (fs/file *log-dir* "01-02-2021-1.time-log"))
           (str (log/find-previous-log "01-02-2021"))))))

(deftest date-test
  (testing "invalid dates"
    (is (nil? (re-find log/date-re "1-1-2000")))
    (is (nil? (re-find log/date-re "32-01-2000")))
    (is (nil? (re-find log/date-re "12-13-2000")))
    (is (nil? (re-find log/date-re "00-00-0000"))))

  (testing "leap years"
    (doseq [year (range 1900 2400 4)]
      (is (re-find log/date-re (str "29-02-" year))))))

(def log
  [["programming"
    {:descr "time-control",
     :start #inst "2021-03-28T07:05:14.488-00:00",
     :end #inst "2021-03-28T07:15:16.451-00:00"}]
   ["programming"
    {:descr "time-control",
     :start #inst "2021-03-28T07:17:58.789-00:00",
     :end #inst "2021-03-28T07:19:43.537-00:00"}]
   ["home"
    {:descr "housekeeping",
     :start #inst "2021-03-28T08:20:42.926-00:00",
     :end #inst "2021-03-28T08:21:52.395-00:00"}]
   ["home"
    {:descr "cleaning",
     :start #inst "2021-03-28T08:22:01.431-00:00",
     :end #inst "2021-03-28T08:41:39.222-00:00"}]
   ["programming"
    {:descr "time-control",
     :start #inst "2021-03-28T08:41:39.222-00:00",
     :end #inst "2021-03-28T08:41:52.204-00:00"}]
   ["programming"
    {:descr "time-control tests",
     :start #inst "2021-03-28T08:42:59.636-00:00",
     :end #inst "2021-03-28T08:43:48.782-00:00"}]
   ["programming"
    {:descr "learning clojure",
     :start #inst "2021-03-28T08:43:57.638-00:00",
     :end #inst "2021-03-28T09:01:37.087-00:00"}]])

(deftest log-stats
  (testing "category filtering"
    (is (= [["time-control"
             {:start #inst "2021-03-28T07:05:14.488-00:00",
              :end #inst "2021-03-28T07:15:16.451-00:00"}]
            ["time-control"
             {:start #inst "2021-03-28T07:17:58.789-00:00",
              :end #inst "2021-03-28T07:19:43.537-00:00"}]
            ["time-control"
             {:start #inst "2021-03-28T08:41:39.222-00:00",
              :end #inst "2021-03-28T08:41:52.204-00:00"}]
            ["time-control tests"
             {:start #inst "2021-03-28T08:42:59.636-00:00",
              :end #inst "2021-03-28T08:43:48.782-00:00"}]
            ["learning clojure"
             {:start #inst "2021-03-28T08:43:57.638-00:00",
              :end #inst "2021-03-28T09:01:37.087-00:00"}]]
           (#'log/filter-by-category log "programming")))
    (is (= [["housekeeping"
             {:start #inst "2021-03-28T08:20:42.926-00:00",
              :end #inst "2021-03-28T08:21:52.395-00:00"}]
            ["cleaning"
             {:start #inst "2021-03-28T08:22:01.431-00:00",
              :end #inst "2021-03-28T08:41:39.222-00:00"}]]
           (#'log/filter-by-category log "home")))
    (is (= []
           (#'log/filter-by-category log "non existing"))))

  (testing "log stats by category"
    (is (= {"time-control" {:days 0, :hours 0, :minutes 11, :seconds 59},
            "time-control tests" {:days 0, :hours 0, :minutes 0, :seconds 49},
            "learning clojure" {:days 0, :hours 0, :minutes 17, :seconds 39}}
           (#'log/log-stats (#'log/filter-by-category log "programming"))))
    (is (= {"housekeeping" {:days 0, :hours 0, :minutes 1, :seconds 9},
            "cleaning" {:days 0, :hours 0, :minutes 19, :seconds 37}}
           (#'log/log-stats (#'log/filter-by-category log "home")))))

  (testing "total log stats"
    (is (= {"programming" {:days 0, :hours 0, :minutes 30, :seconds 28},
            "home" {:days 0, :hours 0, :minutes 20, :seconds 47}}
           (#'log/log-stats log)))))


test/time_control/user_activities_test.clj
(ns time-control.user-activities-test
  (:require [clojure.java.io :as io]
            [clojure.pprint :as pp]
            [clojure.test :refer [deftest is testing use-fixtures]]
            [me.raynes.fs :as fs]
            [time-control.user-activities :as acts]
            [time-control.log :refer [*log-dir*]]))

(def test-activities {"a" "anime"
                      "h" "housework"
                      "e" "exercise"
                      "t" "training"})


(defn- wrap-activity-file [test]
  (binding [*log-dir* (fs/file "test" "data" "time-log")
            acts/*activities-file* (fs/file "test" "data" "time-log" ".commands.edn")]
    (fs/mkdirs *log-dir*)
    (spit acts/*activities-file* (with-out-str (pp/pprint test-activities)))
    (test)
    (fs/delete acts/*activities-file*)
    (fs/delete-dir *log-dir*)))


(defn- with-logging-slurp [test]
  (with-redefs [slurp (fn [f]
                        (print "reading file")
                        (let [sw (java.io.StringWriter.)]
                          (with-open [^java.io.Reader r (io/reader f)]
                            (io/copy r sw)
                            (.toString sw))))]
    (test)))


(use-fixtures :each
  wrap-activity-file
  with-logging-slurp)


(deftest caching
  (testing "file is not read from disk twice"
    (is (= "reading file"
           (with-out-str (acts/user-activities))))
    (is (= "" (with-out-str (acts/user-activities)))))

  (testing "file is read upon modification"
    (is (= "" (with-out-str (acts/user-activities))))
    (spit acts/*activities-file* "{}")
    (is (= "reading file"
           (with-out-str (acts/user-activities))))
    (is (= {} (acts/user-activities)))))


(deftest adding-removing-activities
  (testing "adding activities"
    (is (= (assoc test-activities "1" "2")
           (do (acts/add-activity ["1" "2"])
               (acts/user-activities)))))

  (testing "removing activities"
    (is (= (dissoc test-activities "1")
           (do (acts/remove-activity ["1"])
               (acts/user-activities))))))


(deftest activity-printing
  (testing "listing activities"
    (is (= "reading filea - anime\ne - exercise\nh - housework\nt - training\n"
           (with-out-str (acts/list-activities))))))


project.clj
(defproject time-control "0.1.0-SNAPSHOT"
  :description "Simple time logging utility"
  :url "https://gitlab.com/andreyorst/time-control"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :plugins [[io.taylorwood/lein-native-image "0.3.1"]]
  :dependencies [[org.clojure/clojure "1.10.3"]
                 [clj-commons/fs "1.6.307"]
                 [org.clojure/tools.cli "1.0.206"]
                 [org.flatland/ordered "1.5.9"]]
  :main ^:skip-aot time-control.core
  :target-path "target/%s"
  :native-image {:name      "time-control"
                 :graal-bin "/home/andreyorst/Downloads/graalvm-ce-java11-21.0.0.2/bin"
                 :opts      ["--verbose"
                             "--no-fallback"
                             "--initialize-at-build-time"
                             "-H:+ReportExceptionStackTraces"
                             "--report-unsupported-elements-at-runtime"
                             ]}
  :profiles {:uberjar {:aot :all
                       :native-image {:jvm-opts ["-Dclojure.compiler.direct-linking=true"]}}})


src/time_control/core.clj
(ns time-control.core
  (:require
   [clojure.edn :as edn]
   [clojure.pprint]
   [clojure.string :as str]
   [clojure.tools.cli :refer [parse-opts]]
   [me.raynes.fs :as fs]
   [time-control.log :refer [create-new-log
                             detailed-log-summary
                             find-previous-log
                             last-log-item-summary
                             log
                             log-activity
                             log-file
                             update-last-log-entry
                             write-current
                             *log-dir*]]
   [time-control.user-activities :refer [add-activity
                                         remove-activity
                                         user-activities
                                         list-activities]])
  (:gen-class))

(set! *warn-on-reflection* true)

(def ^:dynamic *print-interval* 3000000)

(defonce printer (atom nil))
(defonce saver (atom nil))


(defmacro periodic-job
  "Start background job `body` that runs every `period` milliseconds.
  Stores the future in the given `atom`, and cancels previous future
  if the `atom` contains one."
  {:style/indent 2}
  [atom period & body]
  `(let [atom# (deref ~atom)]
     (when (future? atom#)
       (future-cancel atom#))
     (reset! ~atom (future
                     (while (not (Thread/interrupted))
                       (Thread/sleep ~period)
                       ~@body)))))


(defmacro with-term-prompt
  "Wraps the `body` in a way that helps maintaining visual look of the
  log in the terminal."
  [& body]
  `(do (println)
       ~@body
       (print "> ")
       (flush)))


(defn read-input []
  (print "> ")
  (flush)
  (read-line))


(defn quit
  "Cancel any running background jobs, write log file and exit the
  program."
  []
  (when-some [printer @printer]
    (future-cancel printer))
  (when-some [saver @saver]
    (future-cancel saver))
  (write-current)
  (System/exit 0))


;; Each command contains a description `:descr`, and function `:func`
;; to run.  Optional `:args`, `:req-args`, and `:opt-args` are used to
;; enforce minimal correct amount of arguments, and build help message.
(def commands
  {":h" {:descr "Help for available commands"
         :func (fn [& _]
                 (doseq [[key {:keys [descr req-args opt-args]}] commands]
                   (println (str key
                                 (if req-args
                                   (str " <" (str/join "> <" req-args) ">")
                                   "")
                                 (if opt-args
                                   (str " [" (str/join "] [" opt-args) "]")
                                   "")
                                 " - " descr))))}

   ":n" {:descr "Start new log"
         :func (fn [& _]
                 (if-let [log-file @log-file]
                   (println "Log is already selected:" (str log-file))
                   (let [file (create-new-log)]
                     (println "Created new log:" (str file))
                     (reset! log-file file)
                     (log-activity "unknown" "unknown activity")
                     (periodic-job printer *print-interval*
                                   (with-term-prompt (last-log-item-summary))))))}

   ":p" {:descr "Continue previous log"
         :func (fn [& _]
                 (if-let [log-file @log-file]
                   (println "Log is already selected:" (str log-file))
                   (let [file (find-previous-log)]
                     (println "Selected log:" (str file))
                     (reset! log-file file)
                     (reset! log (edn/read-string (slurp file)))
                     (log-activity "unknown" "unknown activity")
                     (periodic-job printer *print-interval*
                                   (with-term-prompt (last-log-item-summary))))))}

   ":c" {:descr "Current activity stats"
         :func (fn [& _] (last-log-item-summary))}

   ":e" {:descr "Exit"
         :func (constantly :exit)}

   ":a" {:descr "Add new activity type"
         :req-args ["name" "description*"]
         :func add-activity}

   ":r" {:descr "Remove activity type"
         :req-args ["name"]
         :func remove-activity}

   ":u" {:descr "Update last activity description and type"
         :req-args ["activity" "description*"]
         :func update-last-log-entry}

   ":l" {:descr "List all available activities"
         :func (fn [& _] (list-activities))}

   ":s" {:descr "Summary for given category and period"
         :req-args ["category"]
         :opt-args ["date|[start-date]:[end-date]|all"]
         :func detailed-log-summary}})


(defn execute
  "Execute `cmd` with `args`.
  If `cmd` starts with colon, and is available in the `commands` map,
  execute respective `:func` wiht `args`.

  If `cmd` is not available in the `commands` map, looks it as an
  activity in `user-activities`, and starts new activity if found.

  Otherwise prints error."
  [[cmd & args]]
  (if-let [command (get commands cmd)]
    (let [{:keys [func req-args]} command]
      (if (or (nil? req-args) (>= (count args) (count req-args)))
        (func args)
        (println (str "Wrong argument amount: " (count args)
                      " for " cmd " (" (count req-args) ")"))))
    (if @log-file
      (if-let [activity (get (user-activities) cmd)]
        (do (log-activity activity (str/join " " args))
            (periodic-job printer *print-interval*
                          (with-term-prompt (last-log-item-summary))))
        (when (not-empty cmd)
          (if (str/starts-with? cmd ":")
            (do (println (str "Unknown command '" cmd "'"))
                (println ":h - list all commands"))
            (do (println (str "unknown activity '" cmd "'"))
                (println ":l - list all activities")))))
      (do (println "No log active")
          (println ":n - create new log")
          (println ":p - continue last log")))))


(defn run
  "Starts two background jobs for logging current activity, and saving
  log file, and one foreground job for user input."
  [{:keys [save-interval]}]
  (periodic-job saver save-interval
                (write-current))
  (loop []
    (if-some [input (read-input)]
      (case (-> input
                (str/split #"\s")
                execute)
        :exit (quit)
        (recur))
      (quit))))


(def arg-opts
  [["-p"
    "--print-interval TIME"
    "Interval in minutes to report current activity status"
    :default 300000
    :default-desc "5"
    :parse-fn (fn [x]
                (-> x
                    (edn/read-string)
                    (* 60 1000)))]
   ["-s"
    "--save-interval TIME"
    "Interval in minutes to save activity log"
    :default 60000
    :default-desc "1"
    :parse-fn (fn [x]
                (-> x
                    (edn/read-string)
                    (* 60 1000)))]
   [nil
    "--log-dir PATH"
    "Path to the log directory"
    :default (str (fs/file (fs/home) "time-log"))
    :default-desc "~/time-log"
    :parse-fn #(str (fs/file (fs/expand-home %)))
    :validate [#(not (fs/file? %))
               "File with this name already exists"]]
   ["-h" "--help" "Print this message and exit"]])


(defn -main
  [& args]
  (let [{:keys [options errors summary]} (parse-opts args arg-opts)
        {:keys [log-dir help]} options]
    (when errors
      (doseq [error errors]
        (println error))
      (System/exit 1))
    (when help
      (println summary)
      (System/exit 0))
    (binding [*log-dir* log-dir
              *print-interval* (:print-interval options)]
      (when-not (fs/exists? *log-dir*)
        (fs/mkdir *log-dir*))
      (run options))))


src/time_control/java_time.clj
(ns time-control.java-time
  (:refer-clojure :exclude [format])
  (:import [java.util Date]
           [java.time LocalDate LocalDateTime Instant ZoneId]
           [java.time.format DateTimeFormatter]))

(set! *warn-on-reflection* true)

(defn format [format-str ^LocalDate date]
  (.format (DateTimeFormatter/ofPattern format-str) date))


(defn local-date ^LocalDate
  ([] (LocalDate/now))
  ([^Date date]
   (LocalDate/ofInstant (.toInstant date) (ZoneId/systemDefault)))
  ([format-str date-str]
   (LocalDate/parse date-str (DateTimeFormatter/ofPattern format-str))))


(defn before? [^LocalDate a ^LocalDate b]
  (.isBefore a b))


(defn after? [^LocalDate a ^LocalDate b]
  (.isAfter a b))


(defn local-date-time ^LocalDate
  ([]
   (LocalDateTime/now))
  ([^Date date]
   (LocalDateTime/ofInstant (.toInstant date) (ZoneId/systemDefault)))
  ([format-str date-str]
   (LocalDateTime/parse date-str (DateTimeFormatter/ofPattern format-str))))


(defn now ^Date []
  (Date/from (Instant/now)))


src/time_control/user_activities.clj
(ns time-control.user-activities
  (:require [clojure.edn :as edn]
            [clojure.pprint :as pp]
            [clojure.string :as str]
            [me.raynes.fs :as fs]))

(set! *warn-on-reflection* true)

(def ^:dynamic *activities-file* (str (fs/file (fs/home) "time-log" ".commands.edn")))


;; Cached read of activity config file.  If file modification time is
;; newer than modification time stored in the cache, re reads
;; information from the disk.  Otherwise uses data from the cache, if
;; available. Note, this may look like a variable, but this is a
;; function, that is created with the #() notation.
(def user-activities
  (let [cache (atom {:mod-time 0
                     :config {}})]
    #(let [mod-time (fs/mod-time *activities-file*)]
       (when (> mod-time (:mod-time @cache))
         (swap! cache assoc
                :config (edn/read-string (slurp *activities-file*))
                :mod-time mod-time))
       (:config @cache))))


(defn add-activity
  "Adds activity `name` and `descriotions` to the activity config."
  [[name & descriptions]]
  (let [acts (assoc (user-activities) name (str/join " " descriptions))]
    (spit *activities-file* (with-out-str (pp/pprint acts)))))


(defn remove-activity
  "Removes activity `name` from the activity config."
  [[name]]
  (let [acts (dissoc (user-activities) name)]
    (spit *activities-file* (with-out-str (pp/pprint acts)))))


(defn list-activities
  "Prints all available activities."
  [& _]
  (if-let [activities (seq (sort-by first compare (user-activities)))]
    (doseq [[activity descr] activities]
      (println activity "-" descr))
    (println "No activities are defined. Use :a to create new activity type")))


src/time_control/log.clj
(ns time-control.log
  (:require [clojure.edn :as edn]
            [clojure.pprint :as pp]
            [clojure.string :as str]
            [me.raynes.fs :as fs]
            [time-control.user-activities :refer [user-activities]]
            [time-control.java-time :as jt]
            [flatland.ordered.map :refer [ordered-map]])
  (:import [java.util Date]))

(set! *warn-on-reflection* true)

(def ^:dynamic *log-dir* (str (fs/file (fs/home) "time-log")))

(defonce log-file (atom nil))
(defonce log (atom []))

(def ^:const date-re
  (re-pattern (str #"(?:(?:(?:0[1-9]|1[0-9]|2[0-8])-(?:0[1-9]|1[012]))|(?:29|30|31)-(?:0[13578]|1[02])|(?:29|30)-(?:0[4,6,9]|11))-(?:19|[2-9][0-9])\d\d"
                   #"|(?:29-02-(?:19|[2-9][0-9])(?:00|04|08|12|16|20|24|28|32|36|40|44|48|52|56|60|64|68|72|76|80|84|88|92|96))")))

(defn- extract-log-date-id
  "Extracts date string and optional index from `file`."
  [file]
  (->> file
       str
       (re-find (re-pattern (str "(" date-re ")(?:-(\\d+))?")))
       (drop 1)))


(defn- compare-log-names
  "Comparison of two log names based on dates and optional indexes."
  [[date1 num1] [date2 num2]]
  (or (jt/before? (jt/local-date "dd-MM-yyyy" date1)
                  (jt/local-date "dd-MM-yyyy" date2))
      (and (= date1 date2)
           (< (if num1 (Integer/parseInt num1) 0)
              (if num2 (Integer/parseInt num2) 0)))))


(defn- get-sorted-logs
  "Finds all logs, or logs that start with `name` in the `*log-dir` and
  sorts those by date."
  ([] (get-sorted-logs nil))
  ([date]
   (let [glob (if (some? date)
                (str date "-*.time-log")
                "*.time-log")]
     (some->> glob
              (fs/glob (fs/file *log-dir*))
              (sort-by extract-log-date-id compare-log-names)))))


(defn find-previous-log
  "Get last log by date."
  ([] (last (get-sorted-logs)))
  ([date] (last (get-sorted-logs date))))


(defn- get-log-name-index
  "For logs that contain index in the name returns that index."
  [file]
  (or (some->> file
               str
               (re-find (re-pattern (str "(?:" date-re ")-(\\d+)\\.time-log$")))
               second
               Integer/parseInt)
      0))


(defn create-new-log
  "Creates new log file."
  ([] (create-new-log (jt/format "dd-MM-yyyy" (jt/local-date))))
  ([date]
   (let [filename (str date ".time-log")
         unique-filename (if (fs/exists? (fs/file *log-dir* filename))
                           (let [prev (find-previous-log date)]
                             (str date "-" (inc (get-log-name-index prev)) ".time-log"))
                           filename)
         log-file (fs/file *log-dir* unique-filename)]
     (doto log-file
       (spit "[]")))))


(defn- close-last-item
  "If the last item in the log doesn't have `:end` date, adds current
  date to it."
  [log]
  (let [last (dec (count log))]
    (if (and (>= last 0)
             (nth log last)
             (nil? (get-in log [last 1 :end])))
      (assoc-in log [last 1 :end] (jt/now))
      log)))


(defn- millis-to-dhms
  "Convert `millis` to days, hours, minutes, and seconds."
  [millis]
  {:days (int (/ millis 1000 60 60 24))
   :hours (int (mod (/ millis 1000 60 60) 24))
   :minutes (int (mod (/ millis 1000 60) 60))
   :seconds (int (mod (/ millis 1000) 60))})


(defn- millis-between
  "Gets amount of milliseconds between two dates."
  [date-a date-b]
  (let [res (- (inst-ms date-a) (inst-ms date-b))]
    (if (< res 0) (- res) res)))


(defn- reduce-log-times
  "Reducing function for calculating total time for given `category`."
  [total-stats [category {:keys [start end]}]]
  (if (get total-stats category)
    (update total-stats category (fn [now start end]
                                   (+ now (millis-between start (or end (jt/now)))))
            start end)
    (assoc total-stats category (millis-between start (or end (jt/now))))))


(defn- log-stats
  "Full log stats."
  [log]
  (some->> log
           (reduce reduce-log-times (ordered-map))
           (map (fn [[k v]] [k {:dhms (millis-to-dhms v) :ms v}]))
           (into (ordered-map))))

(defn- beg-end-log-dates
  "Find the earliest and latest dates in the log."
  [log]
  (loop [^Date log-start nil
         ^Date log-end nil
         lines log]
    (if lines
      (let [{:keys [^Date start ^Date end]} (second (first lines))]
        (recur (if log-start
                 (if (.before start log-start) start log-start)
                 start)
               (let [^Date end (or end (jt/now))]
                 (if log-end
                   (if (.after end log-end) end log-end)
                   end))
               (next lines)))
      {:start log-start
       :end log-end})))


(defn- time-summary-str
  "Prints formatted message about time spent for given category."
  [{:keys [days hours minutes seconds]
    :or {days 0 hours 0 minutes 0 seconds 0}}]
  (str/trim (cond-> ""
              (> days 0) (str days "d ")
              (> hours 0) (str hours "h ")
              (> minutes 0) (str minutes "m ")
              (> seconds 0) (str seconds "s "))))


(defn- print-log-start-end [log]
  (let [{:keys [^Date start ^Date end]} (beg-end-log-dates log)
        start (jt/local-date-time start)
        end (jt/local-date-time end)
        [start-date start-time] (str/split (jt/format "dd-MM-yyyy HH:mm:ss" start) #" ")
        [end-date end-time] (str/split (jt/format "dd-MM-yyyy HH:mm:ss" end) #" ")]
    (println (if (= start-date end-date)
               (str "Activity summary for " start-date
                    " from " start-time " to " end-time)
               (str "Activity summary from "
                    start-date " " start-time
                    " to "
                    end-date " " end-time)))))


(defn- log-summary-str
  "Prints time statistics for the given log."
  [log]
  (if (seq log)
    (let [total-stats (log-stats log)
          total-ms (reduce (fn [t [_ {ms :ms}]] (+ t ms)) 0 total-stats)]
      (->> total-stats
           (mapv (fn [[category {:keys [dhms ms]}]]
                   {"category" category
                    "%" (/ (int (* 100 (/ (* ms 100) total-ms))) 100.0)
                    "time" (time-summary-str dhms)}))
           (sort-by #(get % "%") >)
           (mapv (fn [row] (update row "%" #(str % "%"))))
           (pp/print-table ["category" "%" "time"])
           with-out-str
           str/split-lines
           (drop 3)
           (str/join "\n")))
    (when @log-file
      "No data")))


(defn log-summary
  "Public API for printing activity stats only for the current log."
  []
  (let [log @log]
    (print-log-start-end log)
    (println (log-summary-str log))))


(defn last-log-item-summary
  "Prints activity summary for last log item."
  []
  (let [log @log]
    (if (seq log)
      (let [[category {:keys [descr]} :as last-line] (last log)
            row-stats (log-stats [last-line])]
        (println (if-not (empty? descr)
                   (str category " - " descr ":")
                   (str category ":"))

                 (time-summary-str (:dhms (first (vals row-stats))))))
      (when @log-file
        (println "No data")))))


(defn log-activity
  "Adds new activity to the log.  Closes previous activity if not
  closed."
  [type descr]
  (println "Started new activity:"
           (if-not (empty? descr) (str type " - " descr) type))
  (swap! log close-last-item)
  (swap! log conj [type {:descr descr
                         :start (jt/now)}]))

(defn write-current
  "Store current log on the file system."
  []
  (when-some [log-file @log-file]
    (some->> @log
             close-last-item
             (#(with-out-str (pp/pprint %)))
             (spit log-file))))


(defn update-last-log-entry
  "Change the category and description of the last log item."
  [[category & descriptions]]
  (let [l @log
        last (dec (count l))]
    (when-let [line (and (>= last 0) (nth l last))]
      (println "Updated last activity to"
               (if-not (empty? category)
                 (str category " - " (str/join " " descriptions))
                 category))
      (swap! log assoc last (-> line
                                (assoc-in [0] category)
                                (assoc-in [1 :descr] (str/join " " descriptions)))))))


(defn- gather-logs-in-period
  "Get all log files for the given `period`.  If the `period` is `all`
  gathers all logs in the `*log-dir*`.  If the `period` is date in the
  format of `dd-MM-yyyy` gathers the logs for this day only.  If the
  `period` is two dates separated with colon, gathers logs for dates
  in this period.  Otherwise gets all logs for current day."
  [period]
  (try
    (let [period (or period "")]
      (cond (or (= period "all")
                (= period ":"))
            (into []
                  (comp (map #(edn/read-string (slurp %)))
                        cat)
                  (get-sorted-logs))

            (re-find #":" period)
            (let  [[start end] (str/split period #":")
                   end (jt/local-date "dd-MM-yyyy" (if (empty? end) "01-01-3000" end))
                   start (jt/local-date "dd-MM-yyyy" (if (empty? start) "01-01-1900" start))
                   logs (get-sorted-logs)]
              (into []
                    (comp (filter #(let [[date] (extract-log-date-id %)]
                                     (not (jt/before? (jt/local-date "dd-MM-yyyy" date) start))))
                          (filter #(let [[date] (extract-log-date-id %)]
                                     (not (jt/after? (jt/local-date "dd-MM-yyyy" date) end))))
                          (map #(edn/read-string (slurp %)))
                          cat)
                    logs))
            (re-find date-re period)
            (gather-logs-in-period (str period ":" period))
            (nil? period)
            (gather-logs-in-period (jt/format "dd-MM-yyyy" (jt/local-date)))

            :else
            (println "Unsupported period. Use empty period, date, beg-date:end-date, or all")))
    (catch Exception e
      (println e "Error during reading logs:" (.getMessage e)))))


(defn- filter-by-category [log category]
  (sequence (comp (keep #(when (= category (first %)) %))
                  (map #(let [m (second %)
                              category (:descr m)
                              m (dissoc m :descr)]
                          [category m])))
            log))


(defn- detailed-log-summary' [log category]
  (if-let [category-log (seq (filter-by-category log category))]
    (let [{:keys [days hours minutes seconds]
           :or {days 0 hours 0 minutes 0 seconds 0}
           :as dhms}
          (:dhms (get (log-stats log) category))]
      (println)
      (if (pos-int? (+ days hours minutes seconds))
        (println (str category " total time: " (time-summary-str dhms)))
        (println category))
      (println (log-summary-str category-log)))
    (println "No stats for category" category)))


(defn detailed-log-summary
  "Prints detailed log summary for given `category` and `period`."
  [[category period]]
  (let [period (or period (jt/format "dd-MM-yyyy" (jt/local-date)))]
    (if-let [logs (seq (gather-logs-in-period period))]
      (do (print-log-start-end logs)
          (case category
            "all" (doseq [category (vals (user-activities))]
                    (detailed-log-summary' logs category))
            (detailed-log-summary' logs category)))
      (println "No stats for period" period))))
